exporter/prometheusremotewriteexporter/wal.go

```diff
@@ -0,0 +1,246 @@
+	if wc == nil {
+		// There are cases for which the WAL can be disabled.
+		// TODO: Perhaps log that the WAL wasn't enabled.
+		return nil, nil
+	}
+	wal, err := wal.Open(filepath.Join(wc.Dir, "prom_rw"), &wal.Options{
+		SegmentCacheSize: wc.nBeforeTruncation(),
+		NoCopy:           true,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("prometheusremotewriteexporter: failed to open WAL: %w", err)
+	}
+	return wal, nil
+}
+
+func (wc *walConfig) nBeforeTruncation() int {
+	if wc == nil {
+		return 0
+	}
+	if wc.NBeforeTruncation <= 0 {
+		return 300
+	}
+	return wc.NBeforeTruncation
+}
+
+func (wc *walConfig) refreshDuration() time.Duration {
+	if wc == nil {
+		return 0
+	}
+	if wc.RefreshDuration <= 0 {
+		// Following what Prometheus's WAL defaults.
+		return 1 * time.Minute
+	}
+	return wc.RefreshDuration
+}
+
+func (prwe *PrwExporter) walEnabled() bool { return prwe.walConfig != nil }
+
+var errWALDisabled = errors.New("WAL is disabled")
+
+func (prwe *PrwExporter) retrieveWALIndices(context.Context) (err error) {
+	if !prwe.walEnabled() {
+		return errWALDisabled
+	}
+
+	prwe.walMu.Lock()
+	defer prwe.walMu.Unlock()
+
+	wal, err := prwe.walConfig.createWAL()
+	if err != nil {
+		return err
+	}
+	prwe.wal = wal
+
+	prwe.rWALIndex, err = prwe.wal.FirstIndex()
+	if err != nil {
+		return fmt.Errorf("prometheusremotewriteexporter: failed to retrieve the first WAL index: %w", err)
+	}
+	prwe.wWALIndex, err = prwe.wal.LastIndex()
+	if err != nil {
+		return fmt.Errorf("prometheusremotewriteexporter: failed to retrieve the last WAL index: %w", err)
+	}
+	return nil
+}
+
+// exportThenTruncateWAL uploads the serialized data to the remote write endpoint,
+// and then truncates and retrieves the new indices for the WAL.
+func (prwe *PrwExporter) exportThenTruncateWAL(ctx context.Context, reqL []*prompb.WriteRequest) error {
+	if len(reqL) == 0 {
+		return nil
+	}
+	// Otherwise, time to flush.
+	if errs := prwe.exportWriteRequests(ctx, reqL); len(errs) != 0 {
+		return consumererror.Combine(errs)
+	}
+	// Save all the entries that aren't yet committed, to the tail of the WAL.
+	if err := prwe.wal.Sync(); err != nil {
+		return err
+	}
+	// Truncate the WAL from the front for the entries that we already
+	// read from the WAL and then exported.
+	if err := prwe.wal.TruncateFront(prwe.rWALIndex); err != nil && err != wal.ErrOutOfRange {
+		return err
+	}
+	// Reset by retrieving the respective read and write WAL indices.
+	return prwe.retrieveWALIndices(ctx)
+}
+
+func (prwe *PrwExporter) turnOnWALIfEnabled() error {
+	if !prwe.walEnabled() {
+		return nil
+	}
+
+	// From this point down below, the Write-Ahead-Log is enabled.
+	shutdownCtx, cancel := context.WithCancel(context.Background())
+	if err := prwe.retrieveWALIndices(shutdownCtx); err != nil {
+		cancel()
+		return err
+	}
+
+	go func() {
+		<-prwe.closeChan
+		cancel()
+	}()
+
+	prwe.wg.Add(1)
+	go func() {
+		defer prwe.wg.Done()
+
+		_ = prwe.readFromWALThenExport(shutdownCtx)
+	}()
```

Comment:
Why different go-routine? `turnOnWALIfEnabled` should completely finish when this returns correct? Otherwise things like `shutdownCtx` may be already closed.

Reply:
@odeke-em can you respond to Bogdan's questions? Thx.

Reply:
No, readFromWALThenExport will concurrently run throughout the lifetime of the program,
until shutdownCtx is canceled. turnOnWALIfEnabled just triggers the code that'll run until shutdown.

>Otherwise things like shutdownCtx may be already closed.

If it is canceled, readFromWALThenExport will terminate.

Reply:
A component should be entirely shutdown (complete) when the Shutdown returns and should not have any other work scheduled for later. We want every component to be "restartable" and this will make it impossible to do that.

See https://github.com/open-telemetry/opentelemetry-collector/blob/main/component/component.go#L60